bundles/org.openhab.binding.fmiweather/src/main/java/org/openhab/binding/fmiweather/internal/discovery/FMIWeatherDiscoveryService.java

```diff
@@ -0,0 +1,265 @@
+                    .build();
+            thingDiscovered(discoveryResult);
+        });
+        if (logger.isDebugEnabled()) {
+            logger.debug("Filtered cities for forecasts: {}",
+                    filteredCities.stream().map(station -> String.format("%s (%s)", station.name, station.id))
+                            .collect(Collectors.toCollection(TreeSet<String>::new)));
+        }
+    }
+
+    private void createObservationsForStations(@Nullable PointType location) {
+        List<Location> candidateStations = new LinkedList<>();
+        List<Location> filteredStations = new LinkedList<>();
+        cachedStations().peek(station -> {
+            if (logger.isDebugEnabled()) {
+                candidateStations.add(station);
+            }
+        }).filter(location2 -> isClose(location, location2)).peek(station -> {
+            if (logger.isDebugEnabled()) {
+                filteredStations.add(station);
+            }
+        }).forEach(station -> {
+            Map<String, Object> properties = new HashMap<>();
+            properties.put(BindingConstants.FMISID, station.id);
+            DiscoveryResult discoveryResult = DiscoveryResultBuilder
+                    .create(new ThingUID(THING_TYPE_OBSERVATION,
+                            cleanId(String.format("station_%s_%s", station.id, station.name))))
+                    .withLabel(String.format("FMI weather observation for %s", station.name)).withProperties(properties)
+                    .build();
+            thingDiscovered(discoveryResult);
+        });
+        if (logger.isDebugEnabled()) {
+            logger.debug("Candidate stations: {}",
+                    candidateStations.stream().map(station -> String.format("%s (%s)", station.name, station.id))
+                            .collect(Collectors.toCollection(TreeSet<String>::new)));
+            logger.debug("Filtered stations: {}",
+                    filteredStations.stream().map(station -> String.format("%s (%s)", station.name, station.id))
+                            .collect(Collectors.toCollection(TreeSet<String>::new)));
+        }
+    }
+
+    private static String cleanId(String id) {
+        return id.replace("ä", "a").replace("ö", "o").replace("å", "a").replace("Ä", "A").replace("Ö", "O")
+                .replace("Å", "a").replaceAll("[^a-zA-Z0-9_]", "_");
+    }
+
+    private static boolean isClose(@Nullable PointType location, Location location2) {
+        return location == null ? true
+                : new PointType(new DecimalType(location2.latitude), new DecimalType(location2.longitude))
+                        .distanceFrom(location).doubleValue() < FIND_STATION_METERS;
+    }
+
+    @Override
+    protected void thingDiscovered(DiscoveryResult discoveryResult) {
+        logger.trace("Discovered: {}", discoveryResult);
+        super.thingDiscovered(discoveryResult);
+    }
+
+    @SuppressWarnings("null")
```

Comment:
Better annotate `Set<Location>` in the cache declaration as NonNull.

Reply:
Unfortunately local variable `stations` can be null still (`ExpiringCache.getValue()` return `@Nullable` value always)

Reply:
If it can be null, wouldn't it be better to apply a null check and remove the supression?